src/mmds/src/token.rs

```diff
@@ -0,0 +1,547 @@
+        self.check_encryption_count()?;
+        // Create token structure containing the encrypted expiry value.
+        let token_struct = self.create_token_struct(ttl_seconds)?;
+        // Encode struct into base64 in order to obtain token string.
+        let token = TokenAuthority::base64_encode_token_struct(&token_struct)?;
+        // Increase the count of encrypted tokens.
+        self.count += 1;
+
+        Ok(token)
+    }
+
+    /// Create a new Token structure to encrypt.
+    fn create_token_struct(&self, ttl_seconds: u32) -> Result<Token, Error> {
+        // Validate token time to live against bounds.
+        if !TokenAuthority::check_ttl(ttl_seconds) {
+            return Err(Error::InvalidTtlValue(ttl_seconds));
+        }
+
+        // Generate 12-byte random nonce.
+        let mut iv = [0u8; IV_LEN];
+        read_into_buf_from_entropy_pool(&mut iv).map_err(|_| Error::TokenGeneration)?;
+
+        // Compute expiration time from ttl.
+        let expiry = TokenAuthority::compute_expiry(ttl_seconds);
+        // Encrypt expiry using the nonce.
+        let (payload, tag) = self.encrypt_expiry(expiry, iv.as_ref())?;
+
+        Ok(Token::new(iv, payload, tag))
+    }
+
+    /// Encrypt expiry using AES-GCM block cipher and return payload and tag obtained.
+    fn encrypt_expiry(
+        &self,
+        expiry: u64,
+        iv: &[u8],
+    ) -> Result<([u8; PAYLOAD_LEN], [u8; TAG_LEN]), Error> {
+        // Create Nonce object from initialization vector.
+        let nonce = Nonce::from_slice(iv);
+        // Convert expiry u64 value into bytes.
+        let mut expiry_as_bytes: Vec<u8> =
+            bincode::serialize(&expiry).map_err(|_| Error::TokenGeneration)?;
+
+        let tag = self
+            .cipher
+            .encrypt_in_place_detached(nonce, b"", &mut expiry_as_bytes)
+            .map_err(|_| Error::TokenGeneration)?;
+
+        // Payload must be of size `PAYLOAD_LEN`.
+        let payload: [u8; PAYLOAD_LEN] = expiry_as_bytes
+            .try_into()
+            .map_err(|_| Error::TokenGeneration)?;
+        // Tag must be of size `TAG_LEN`.
+        let tag_bytes: [u8; TAG_LEN] = tag
+            .as_slice()
+            .try_into()
+            .map_err(|_| Error::TokenGeneration)?;
+
+        Ok((payload, tag_bytes))
+    }
+
+    /// Encode token structure into a string using base64 encoding.
+    fn base64_encode_token_struct(token: &Token) -> Result<String, Error> {
+        let token_bytes: Vec<u8> = bincode::serialize(token).map_err(|_| Error::TokenGeneration)?;
+        let mut encoded_token = Vec::new();
+        // Ensure vector has enough space for base64 encoding.
+        encoded_token.resize(get_size_of_base64_encoding(token_bytes.len()), 0);
+        // Encode token structure bytes into base64.
+        let size = base64::encode_config_slice(token_bytes, base64::STANDARD, &mut encoded_token);
+        encoded_token.resize(size, 0);
```

Comment:
Is there any advantage in performing these resizing logic ? Instead of simply calling `base64::encode_config()` ?

Reply:
I updated the code for both the encoding and decoding functions.